.github/workflows/ci.yml

```diff
@@ -209,69 +328,464 @@ jobs:
+          destination: ./src/
+
+      - name: Download ${{ format('src/zcash-gtest{0}', matrix.file_ext) }}  artifact
+        uses: ./.github/actions/gcs-download-cloud-storage
+        with:
+          path-prefix: ${{ github.run_id }}/${{ matrix.name }}
+          source: ${{ format('src/zcash-gtest{0}', matrix.file_ext) }}
+          destination: ./src/
+
+      - name: Download ${{ format('src/zcashd{0}', matrix.file_ext) }} artifact
+        uses: ./.github/actions/gcs-download-cloud-storage
+        with:
+          path-prefix: ${{ github.run_id }}/${{ matrix.name }}
+          source: ${{ format('src/zcashd{0}', matrix.file_ext) }}
+          destination: ./src/
+
+      - name: Download ${{ format('src/zcash-cli{0}', matrix.file_ext) }} artifact
+        uses: ./.github/actions/gcs-download-cloud-storage
+        with:
+          path-prefix: ${{ github.run_id }}/${{ matrix.name }}
+          source: ${{ format('src/zcash-cli{0}', matrix.file_ext) }}
+          destination: ./src/
+
+      - name: Make artifact executable
+        if: runner.os != 'Windows'
+        run: |
+          chmod +x ${{ format('./src/test/test_bitcoin{0}', matrix.file_ext) }}
+          chmod +x ${{ format('./src/bench/bench_bitcoin{0}', matrix.file_ext) }}
+          chmod +x ${{ format('./src/zcashd{0}', matrix.file_ext) }}
+          chmod +x ${{ format('./src/zcash-cli{0}', matrix.file_ext) }}
+          chmod +x ${{ format('./src/zcash-gtest{0}', matrix.file_ext) }}
+          chmod +x ${{ format('./src/zcashd-wallet-tool{0}', matrix.file_ext) }}
+          chmod +x ${{ format('./src/zcash-tx{0}', matrix.file_ext) }}
+
+      - name: Run sec-hard test
+        shell: bash
+        run: python3 ./qa/zcash/full_test_suite.py sec-hard
+        env:
+          HOST: ${{ matrix.host }}
+
+  rpc-depends:
+    name: RPC set up tier ${{ matrix.tier }} platform ${{ matrix.platform }}
+    needs: setup
+    runs-on: ${{ matrix.build_os }}
+    continue-on-error: ${{ matrix.tier != 1 }}
+    strategy:
+      matrix:
+        name: ${{ fromJson(needs.setup.outputs.rpc_test_names) }}
+        include: ${{ fromJson(needs.setup.outputs.rpc_test_matrix) }}
+
+    steps:
+      - name: Get Python dependencies for RPC tests
+        run: |
+          python3 -m venv ./venv
+          source ./venv/bin/activate
+          pip install simplejson zmq asyncio
+
+      - uses: actions/cache/save@v3
+        name: cache-venv
+        with:
+          enableCrossOsArchive: true
+          path: venv
+          key: venv-${{ matrix.name }}
+
+  test-rpc:
+    name: RPC tests tier ${{ matrix.tier }} platform ${{ matrix.platform }} test ${{ matrix.rpc_tests }}
+    needs: [setup, build, rpc-depends]
+    runs-on: ${{ matrix.build_os }}
+    continue-on-error: ${{ matrix.tier != 1 }}
+    strategy:
+      matrix:
+        name: ${{ fromJson(needs.setup.outputs.rpc_test_names) }}
+        rpc_tests : ${{ fromJson(needs.setup.outputs.rpc_tests) }}
```

Comment:
I want to solve this before we merge the RPC tests into Actions CI. Right now, because `needs.setup.outputs.rpc_tests` is a list of RPC test names, the effect of combining `name` with `rpc_tests` is that we create a separate job per RPC test per runner, which I don't believe is actually faster than batching and parallelizing within a runner (particularly because there is an upper limit on the number of jobs we can run at once).

Instead, I think we need to change `needs.setup.outputs.rpc_tests` to be a list of lists of RPC test names, with the code in `setup` deciding how to split the tests into different matrix entry groups.

To avoid blocking this PR, I propose that once we get everything else working, we move the last commit (that adds the RPC tests) to a separate branch, remove it from this one, and then merge this PR to get the rest of the changes in (which we can then remove from Tekton CI).

Reply:
To parallelize within a single runner, would we do it within the Python script?